source/val/validate_mode_setting.cpp

```diff
@@ -68,6 +235,122 @@ spv_result_t ValidateExecutionMode(ValidationState_t& _,
            << "' is not the Entry Point "
               "operand of an OpEntryPoint.";
   }
+
+  const auto mode = inst->GetOperandAs<SpvExecutionMode>(1);
+  const auto* models = _.GetExecutionModels(entry_point_id);
+  switch (mode) {
+    case SpvExecutionModeInvocations:
+    case SpvExecutionModeInputPoints:
+    case SpvExecutionModeInputLines:
+    case SpvExecutionModeInputLinesAdjacency:
+    case SpvExecutionModeInputTrianglesAdjacency:
+    case SpvExecutionModeOutputPoints:
+    case SpvExecutionModeOutputLineStrip:
+    case SpvExecutionModeOutputTriangleStrip:
+      if (!std::all_of(models->begin(), models->end(),
+                       [](const SpvExecutionModel& model) {
+                         return model == SpvExecutionModelGeometry;
+                       })) {
+        return _.diag(SPV_ERROR_INVALID_DATA, inst)
+               << "Execution mode can only be used with the Geometry execution "
+                  "model.";
+      }
+      break;
+    case SpvExecutionModeSpacingEqual:
+    case SpvExecutionModeSpacingFractionalEven:
+    case SpvExecutionModeSpacingFractionalOdd:
+    case SpvExecutionModeVertexOrderCw:
+    case SpvExecutionModeVertexOrderCcw:
+    case SpvExecutionModePointMode:
+    case SpvExecutionModeQuads:
+    case SpvExecutionModeIsolines:
+      if (!std::all_of(
+              models->begin(), models->end(),
+              [](const SpvExecutionModel& model) {
+                return (model == SpvExecutionModelTessellationControl) ||
+                       (model == SpvExecutionModelTessellationEvaluation);
+              })) {
+        return _.diag(SPV_ERROR_INVALID_DATA, inst)
+               << "Execution mode can only be used with a tessellation "
+                  "execution model.";
+      }
+      break;
+    case SpvExecutionModeTriangles:
+    case SpvExecutionModeOutputVertices:
+      if (!std::all_of(models->begin(), models->end(),
+                       [](const SpvExecutionModel& model) {
+                         switch (model) {
+                           case SpvExecutionModelGeometry:
+                           case SpvExecutionModelTessellationControl:
+                           case SpvExecutionModelTessellationEvaluation:
```

Comment:
Do we care about MeshShadingNV for OutputVertices?

Reply:
I added limited validation where it would otherwise fail valid code for MeshNV.